Add ability to import Google docs into posts

Will-Howard opened the pull request:
_This will be against master once https://github.com/ForumMagnum/ForumMagnum/pull/8787 is merged_

Based on these [designs](https://www.figma.com/file/IRUVQ5D2mKxoW0xZnhXIWv/2024-Q1?node-id=520%3A9186&mode=dev). There are a few differences from the designs, some of which are based on [this slack thread](https://cea-core.slack.com/archives/C03G0AQ925P/p1707135346703839), and some of which are things I changed (@agnesstenlund I can change them back if you want).

The changes from the thread are:
- Removing the option to link your account

The additional changes I made:
- I changed the copy a bit, including removing the "Import shared Google doc" header (because this is implied now there is only one option)
- I made it so all the bits of the UI (the button and the "This will overwrite" message) show before you enter a url. I did this because I personally find it a bit annoying when you can't tell what all the steps are before you fill in part of a form (mentioned in the slack thread), and also now there is only one form I think it looks ok to have more info in one go

### Summary of the logic

Clicking "Import Google doc" does the following:
- Exports the doc as html, using a "service account" (that an admin has to log into to set up). For us the account is `eaforum.posts@gmail.com`. I decided to go with a plain gmail account to avoid any risk of exposing CEA files (because you'll be able to download any doc this account has access to), also it's a bit shorter
- Do some conversion (see `convertImportedGoogleDoc`) of the html to make it work with ckeditor, this is mostly to replicate things we do on paste in ckeditor
- Create a new revision and/or post depending on whether this is the new post or edit post form:
  - For new posts, create a draft post with the converted html (which also creates a revision in a callback)
  - For editing posts, create a new _revision_ (this being the latest revision will cause it to load in the editor), but don't set this as the live revision on the post. This means they will need to hit publish to update the live post
- Refresh the page

### Screenshots

![Screenshot 2024-02-20 at 11 44 46](https://github.com/ForumMagnum/ForumMagnum/assets/77623106/b69e4511-9296-4620-8807-0f4629d4b87d)
![Screenshot 2024-02-20 at 11 44 54](https://github.com/ForumMagnum/ForumMagnum/assets/77623106/d8470031-c988-4b77-a284-fc4d80197f3d)
![Screenshot 2024-02-20 at 11 45 01](https://github.com/ForumMagnum/ForumMagnum/assets/77623106/4cad2bf3-8413-4c78-9898-1216ba350cf2)
![Screenshot 2024-02-20 at 11 45 24](https://github.com/ForumMagnum/ForumMagnum/assets/77623106/dd3ef3a5-264e-4228-89d0-d58d48675443)
![Screenshot 2024-02-20 at 11 41 11](https://github.com/ForumMagnum/ForumMagnum/assets/77623106/669fda62-d5c1-4505-b7df-553a58fb42ca)

Things left to do, that I will aim to do before merging:
- [x] Add analytics
- [x] Add a reminder to admins (in the frontpage sidebar) to re-login to the service account when the session gets close to expiring
- [ ] Stretch: Make the doc url pre-fill if you are on a post where you have previously imported from a google doc
- [ ] Stretch: Add tests for the `convertImportedGoogleDoc` function (which replicates the logic in ckeditor)

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1206584836224763) by [Unito](https://www.unito.io)

agnesstenlund commented:
Looks great! Checking now if I wanna make some small UI tweaks, but looks good overall 👍

agnesstenlund commented:
would it be possible to make the "this will overwrite any unsaved changes" conditional on there being any changes? No big deal if not, but would look cleaner when the page is empty

agnesstenlund commented:
anther thing–will it not also overwrite _saved_ changes?

agnesstenlund commented:
it takes 1 sec before the button become non-disabled, any chance that could be faster? It's long enough for me to get confused that it's not working:

https://github.com/ForumMagnum/ForumMagnum/assets/13235378/e369975c-b5dc-46ed-96ce-87991a124777

agnesstenlund commented:
fyi I changed my mind slightly about the colors/ UI when seeing it live–changed to black and grey:
![Screenshot 2024-02-20 at 13 53 16](https://github.com/ForumMagnum/ForumMagnum/assets/13235378/424f051e-05d1-439b-9d91-9baa480aa547)

agnesstenlund commented:
If I delete the text in the input the button doesn't go back to disabled unless I refresh:


https://github.com/ForumMagnum/ForumMagnum/assets/13235378/d78731e2-db85-4645-b6fa-203f4938a84d

Will-Howard commented:
Comments from Lizka:

> Copy changes:
> - [Not exactly related to gdoc import] “Share draft” should IMO be “Share this draft” — next to the gdoc import it looks like it has something to do with gdocs…

Seems good 👍 , I've updated this

> - “Paste a link…” maybe is clearer if we write “Paste a link to your Google Doc here; make sure the anyone with the link can view or the Doc is shared with [email address].” (It’s more words, but less disorienting imo)

Hmm I prefer the original (bc it's fewer words), interested in what @agnesstenlund thinks. I think I'm assuming people will basically have a good understanding of how google doc permissions work so will get the idea without much explanation. Here's a side by side of the two wordings:

<img width="328" alt="Screenshot 2024-02-20 at 14 40 19" src="https://github.com/ForumMagnum/ForumMagnum/assets/77623106/51454781-047d-451b-9ab6-9a180173075e">
<img width="328" alt="Screenshot 2024-02-20 at 14 40 54" src="https://github.com/ForumMagnum/ForumMagnum/assets/77623106/76a9a031-d6cd-45d6-ab97-f7dc609d016e">

> - Love the red “We don’t have access” thing — allays basically all of my concerns
>   - If I go and update the doc and don’t change anything on this tab, will it automatically update? Otherwise I’d add a “Re-upload” button or something if possible

Good catch, I've changed it so it doesn't update automatically. Unfortunately this means we now might be at risk of hitting a rate limit, although according to my calculation it should be able to tolerate ~80 users using it concurrently, which should hopefully be way more than we need

> - I don’t get the “Google Doc import service account” screenshot, but I assume it’s not meant to be viewable to users? If it is, I’d change that page.

Yup this is for admins (and only devs basically), users won't see this

agnesstenlund commented:
Agree that the rewrite is too long. Since we say Google doc above I don't think it's necessary + the length of that line makes me slightly more confused about it?

agnesstenlund commented:
could do something like "Paste a Google doc link here. It needs to be public or shared with eaforum.posts@gmail.com" if we think we should mention google doc

Will-Howard commented:
> could do something like "Paste your Google doc link here. It needs to be public or shared with [eaforum.posts@gmail.com](mailto:eaforum.posts@gmail.com)" if we think we should mention google doc

Eh 🤷‍♂️, I think it's hopefully obvious from the placeholder and button text

Will-Howard commented:
@oetherington btw the error in the console is a (non-serious) bug in the google api library, for which there is a [fix that will be merged soon](https://github.com/googleapis/google-api-nodejs-client/issues/3419)

jpaddison3 commented:
I love that when you link to that fix it puts a backlink in the referenced issue. Like, "we're waiting on you bro".

oetherington commented on packages/lesswrong/server/editor/conversionUtils.ts:
```diff
@@ -412,3 +413,160 @@ export async function dataToWordCount(data: AnyBecauseTodo, type: string) {
+  const { html: withRehostedImages } = await convertImagesInHTML(html, postId, url => url.includes("googleusercontent"))
+  const withConvertedFootnotes = googleDocConvertFootnotes(withRehostedImages)
+  const withNormalisedRedirects = googleDocRemoveRedirects(withConvertedFootnotes)
+  const ckEditorMarkup = await dataToCkEditor(withNormalisedRedirects, "html")
```
A couple of import issues I noticed that probably need to be handled here:

- The document title got imported as normal text at the top of the document, but maybe we should use it to set the post title and remove it from the document body?
- Bold, italics and underline don't work
- Internal links still point to the google doc, maybe we can fix this now or is it still difficult?

Will-Howard replied:
Ok I did the bold/italics one, and made a [ticket](https://app.asana.com/0/0/1206726735026907/f) for internal links

oetherington reviewed (approved):
LGTM.

I left quite a few comments but I don't think anything is necessarily blocking, except maybe admin gating the auth routes (if I'm understanding correctly how those are supposed to work).

Will-Howard commented on package.json:
```diff
@@ -209,6 +209,7 @@
     "express-session": "^1.17.1",
     "feedparser-promised": "^1.2.2",
     "fp-ts": "^2.13.1",
+    "googleapis": "^133.0.0",
```
I'll give them a couple more weeks, but I'll set a reminder with a deadline so we don't just get stuck with this forever

Will-Howard commented on packages/lesswrong/components/posts/GoogleDocImportButton.tsx:
```diff
@@ -0,0 +1,274 @@
+
+  // Re-check the access every 5s, so if the user updates access they won't have to refresh
+  useEffect(() => {
+    let intervalId: NodeJS.Timeout | null = null;
```
Hm the type of `setInterval` says different, anyway I've just made it
```typescript
let intervalId: ReturnType<typeof setInterval> | null = null;
```

Will-Howard commented on packages/lesswrong/server/authenticationMiddlewares.ts:
```diff
@@ -162,6 +165,83 @@ async function deserializeUserPassport(id: AnyBecauseTodo, done: AnyBecauseTodo)
+  const callbackUrl = "google_oauth2callback"
+  const oauth2Client = new OAuth2Client(googleClientId, googleOAuthSecret, `${getSiteUrl()}${callbackUrl}`);
+
+  addConnectHandler('/auth/linkgdrive', (req: Request, res: Response) => {
```
Yes they should! Great catch

darkruby501 commented on packages/lesswrong/server/authenticationMiddlewares.ts:
```diff
@@ -162,6 +165,83 @@ async function deserializeUserPassport(id: AnyBecauseTodo, done: AnyBecauseTodo)
+ * Add routes for handling linking the service account required to import google docs
+ */
+const addGoogleDriveLinkMiddleware = (addConnectHandler: AddMiddlewareType) => {
+  const googleClientId = googleClientIdSetting.get();
+  const googleOAuthSecret = googleOAuthSecretSetting.get();
```
I think I know what you did here, but @Will-Howard, can you elaborate on the kind of credentials you set up?

The part I need to figure out is the consent screen, I think. How did that work for you?
